### PR TITLE
fix: Tolerate malformed workflows when retrying

### PR DESCRIPTION
**Pull request by simster7**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx [hunk 1]
```diff
@@ -263,6 +273,19 @@ export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRe
         this.setState(newChanges);
     }
 
+    private getNode(nodeId: string): NodeStatus {
+        const node: NodeStatus = this.getNode(nodeId);
+        if (!node) {
+            // We may arrive here if we retry a workflow with a failed pod. During the retry operation, the node corresponding
+            // for the failed pod gets deleted, but the child reference of the parent nodes does not, causing them to refer to
+            // a node that doesn't exist. As soon as the workflow gets operated on again, the node will be recreated. For the
+            // time before that happens, don't error out.
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:282]: Another approach to this fix would be to delete the child references in parent nodes when retrying a Workflow... I think I may favor that approach instead. Thoughts?

Comment (alexec) [reply ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:282]: better to fix the underlying data

Comment (simster7) [reply ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:282]: Will do both, I think it's best to leave this in the UI as a stop-gap

Comment (simster7) [reply ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:282]: I added the fix for the underlying data. I should probably remove this comment then?

File: workflow/util/util.go [hunk 2]
```diff
@@ -689,6 +689,7 @@ func RetryWorkflow(kubeClient kubernetes.Interface, hydrator hydrator.Interface,
 	}
 
 	// Iterate the previous nodes. If it was successful Pod carry it forward
+	deletedNodes := make(map[string]bool)
```
Comment (alexec) [inline workflow/util/util.go:692]: this only happens on retry - is something wrong with retry code ?

Comment (simster7) [reply workflow/util/util.go:692]: Retry is the only code that deletes nodes off a Workflow.

File: ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx [hunk 3]
```diff
@@ -178,12 +181,19 @@ export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRe
...
 
+    public componentDidUpdate(prevProps: Readonly<WorkflowDagProps>, prevState: Readonly<WorkflowDagRenderOptions>, snapshot?: any): void {
+        if (prevState.waitForUpdate) {
+            this.setState({waitForUpdate: false});
+        }
+    }
+
     public render() {
-        if (!this.props.nodes) {
+        if (!this.props.nodes || this.state.waitForUpdate) {
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:196]: Probably want to distinguish the UX between "Loading" and "Workflow is malformed, waiting for an update"

File: ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx (hunk 1 above)
Comment (alexec) [reply ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:282]: sure!

File: workflow/util/util.go [hunk 4]
```diff
@@ -731,6 +734,30 @@ func RetryWorkflow(kubeClient kubernetes.Interface, hydrator hydrator.Interface,
 		}
 	}
 
+	if len(deletedNodes) > 0 {
+		for _, node := range newWF.Status.Nodes {
+			if len(node.Children) > 0 {
```
Comment (alexec) [inline workflow/util/util.go:739]: minor: I don't think you need this guard - in `len(children) == 0`, then `newChildren` will be `nil`, so `node.Children` will become/stay nil

File: workflow/util/util.go [hunk 5]
```diff
@@ -731,6 +734,30 @@ func RetryWorkflow(kubeClient kubernetes.Interface, hydrator hydrator.Interface,
...
+				var newChildren []string
+				for _, child := range node.Children {
+					if !deletedNodes[child] {
+						newChildren = append(newChildren, child)
+					}
+				}
+				node.Children = newChildren
+			}
+
+			if len(node.OutboundNodes) > 0 {
```
Comment (alexec) [inline workflow/util/util.go:749]: minor: as above

File: workflow/util/util.go (hunk 4 above)
Comment (simster7) [reply workflow/util/util.go]: Yup! Good point. Done.

File: workflow/util/util.go (hunk 5 above)
Comment (simster7) [reply workflow/util/util.go]: Done.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
Good looking test.

**Comment by simster7** [comment]
Back-ported to 2.11

**Pull request merged**